File: backend/app/modules/schedules/schemas/punch_accounting.py
"""Schémas API comptabilisation pointages."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

SlotDetection = Literal["shift_code", "nearest_entry", "planning_first"]
ReviewStatus = Literal["pending", "approved", "rejected"]
OvertimeReason = Literal["early_entry", "late_exit", "daily_excess"]


class PunchAccountingSettingsResponse(BaseModel):
    configured: bool = False
    enabled: bool = False
    tolerance_minutes: int = 30
    default_break_deduct_minutes: int = 45
    use_last_nonzero_exit: bool = True
    slot_detection: SlotDetection = "shift_code"
    within_tolerance_pay_theoretical: bool = True
    require_manager_validation_for_overtime: bool = True


class PunchAccountingSettingsUpdate(BaseModel):
    enabled: Optional[bool] = None
    tolerance_minutes: Optional[int] = Field(None, ge=0, le=120)
    default_break_deduct_minutes: Optional[int] = Field(None, ge=0, le=180)
    use_last_nonzero_exit: Optional[bool] = None
    slot_detection: Optional[SlotDetection] = None
    within_tolerance_pay_theoretical: Optional[bool] = None
    require_manager_validation_for_overtime: Optional[bool] = None


class PunchShiftSlotResponse(BaseModel):
    id: str
    code: Optional[str] = None
    label: str = ""
    entry_time: str
    exit_time: str
    theoretical_gross_minutes: int = 465
    break_deduct_minutes: int = 45
    paid_lunch_break: bool = False
    sort_order: int = 0


class PunchShiftSlotCreate(BaseModel):
    code: Optional[str] = None
    label: str = ""
    entry_time: str
    exit_time: str
    theoretical_gross_minutes: int = Field(465, ge=1, le=960)
    break_deduct_minutes: int = Field(45, ge=0, le=180)
    paid_lunch_break: bool = False
    sort_order: int = 0


class PunchShiftSlotUpdate(BaseModel):
    code: Optional[str] = None
    label: Optional[str] = None
    entry_time: Optional[str] = None
    exit_time: Optional[str] = None
    theoretical_gross_minutes: Optional[int] = Field(None, ge=1, le=960)
    break_deduct_minutes: Optional[int] = Field(None, ge=0, le=180)
    paid_lunch_break: Optional[bool] = None
    sort_order: Optional[int] = None


class PunchOvertimeReviewResponse(BaseModel):
    id: str
    employee_id: str
    employee_name: Optional[str] = None
    work_date: str
    overtime_hours: float
    reason: OvertimeReason
    raw_entry_time: Optional[str] = None
    raw_exit_time: Optional[str] = None
    applied_slot_id: Optional[str] = None
    status: ReviewStatus
    review_note: Optional[str] = None


class PunchOvertimeReviewUpdate(BaseModel):
    status: ReviewStatus
    review_note: Optional[str] = None

File: frontend/src/api/punchAccounting.ts
import apiClient from '@/api/apiClient';

export type SlotDetection = 'shift_code' | 'nearest_entry' | 'planning_first';
export type PunchReviewStatus = 'pending' | 'approved' | 'rejected';

export interface PunchAccountingSettings {
  configured: boolean;
  enabled: boolean;
  tolerance_minutes: number;
  default_break_deduct_minutes: number;
  use_last_nonzero_exit: boolean;
  slot_detection: SlotDetection;
  within_tolerance_pay_theoretical: boolean;
  require_manager_validation_for_overtime: boolean;
}

export type PunchAccountingSettingsUpdate = Partial<
  Omit<PunchAccountingSettings, 'configured'>
>;

export interface PunchShiftSlot {
  id: string;
  code?: string | null;
  label: string;
  entry_time: string;
  exit_time: string;
  theoretical_gross_minutes: number;
  break_deduct_minutes: number;
  paid_lunch_break: boolean;
  sort_order: number;
}

export interface PunchShiftSlotCreate {
  code?: string | null;
  label?: string;
  entry_time: string;
  exit_time: string;
  theoretical_gross_minutes?: number;
  break_deduct_minutes?: number;
  paid_lunch_break?: boolean;
  sort_order?: number;
}

export interface PunchOvertimeReview {
  id: string;
  employee_id: string;
  employee_name?: string | null;
  work_date: string;
  overtime_hours: number;
  reason: string;
  raw_entry_time?: string | null;
  raw_exit_time?: string | null;
  status: PunchReviewStatus;
  review_note?: string | null;
}

export const getPunchAccountingSettings = async (): Promise<PunchAccountingSettings> => {
  const { data } = await apiClient.get<PunchAccountingSettings>(
    '/api/schedules/punch-accounting/settings',
  );
  return data;
};

export const updatePunchAccountingSettings = async (
  payload: PunchAccountingSettingsUpdate,
): Promise<PunchAccountingSettings> => {
  const { data } = await apiClient.patch<PunchAccountingSettings>(
    '/api/schedules/punch-accounting/settings',
    payload,
  );
  return data;
};

export const applyPunchAccountingPreset = async (
  preset: string,
): Promise<{ settings: PunchAccountingSettings; slots: PunchShiftSlot[] }> => {
  const { data } = await apiClient.post<{ settings: PunchAccountingSettings; slots: PunchShiftSlot[] }>(
    `/api/schedules/punch-accounting/settings/apply-preset/${preset}`,
  );
  return data;
};

export const listPunchShiftSlots = async (): Promise<PunchShiftSlot[]> => {
  const { data } = await apiClient.get<PunchShiftSlot[]>('/api/schedules/punch-accounting/slots');
  return data;
};

export const createPunchShiftSlot = async (
  payload: PunchShiftSlotCreate,
): Promise<PunchShiftSlot> => {
  const { data } = await apiClient.post<PunchShiftSlot>(
    '/api/schedules/punch-accounting/slots',
    payload,
  );
  return data;
};

export const deletePunchShiftSlot = async (slotId: string): Promise<void> => {
  await apiClient.delete(`/api/schedules/punch-accounting/slots/${slotId}`);
};

export const listPunchOvertimeReviews = async (
  year: number,
  month: number,
  status?: PunchReviewStatus,
): Promise<PunchOvertimeReview[]> => {
  const { data } = await apiClient.get<PunchOvertimeReview[]>(
    '/api/schedules/punch-overtime-reviews',
    { params: { year, month, status } },
  );
  return data;
};

export const updatePunchOvertimeReview = async (
  reviewId: string,
  payload: { status: PunchReviewStatus; review_note?: string },
): Promise<PunchOvertimeReview> => {
  const { data } = await apiClient.patch<PunchOvertimeReview>(
    `/api/schedules/punch-overtime-reviews/${reviewId}`,
    payload,
  );
  return data;
};
